include/seqan3/core/detail/strong_type.hpp

```diff
@@ -0,0 +1,546 @@
+                 _skill_incrementable_concept<unnamed_type>
+        //!\endcond
+    {
+        strong_type_ tmp{this->underlying().get()};
+        ++this->underlying().get();
+        return tmp;
+    }
+
+    //!\brief Adds pre-decrement operator to the strong type.
+    constexpr strong_type_ & operator--()
+        //!\cond
+        requires ((skills & skill::decrement) != skill::none) &&
+                 _skill_decrementable_concept<unnamed_type>
+        //!\endcond
+    {
+        --this->underlying().get();
+        return this->underlying();
+    }
+
+    //!\brief Adds post-decrement operator to the strong type.
+    constexpr strong_type_ operator--(int)
+        //!\cond
+        requires ((skills & skill::decrement) != skill::none) &&
+                 _skill_decrementable_concept<unnamed_type>
+        //!\endcond
+    {
+        strong_type_ tmp{this->underlying().get()};
+        --this->underlying().get();
+        return tmp;
+    }
+    //!\}
+
+    /*!\name Conversion operators.
+     * \{
+     */
+
+    //!\brief Adds explicit conversion to it's underlying type.
+    explicit operator unnamed_type() const
+        //!\cond
+        requires ((skills & skill::convert) != skill::none)
+        //!\endcond
+    {
+        return this->underlying().get();
+    }
+    //!}
+
+protected:
+
+    //!\brief Returns a reference to the associated strong_type.
+    constexpr strong_type_ & underlying() noexcept
+    {
+        return static_cast<strong_type_ &>(*this);
+    }
+
+    //!\brief Returns a reference to the associated strong_type.
+    constexpr strong_type_ const & underlying() const noexcept
+    {
+        return static_cast<strong_type_ const &>(*this);
+    }
+};
+
+/*!\brief Base class to define expressive types.
+ * \ingroup core
+ * \tparam unnamed_type The underlying type to make expressive.
+ * \tparam name_tag     A tag, giving the `unnamed_type` a meaningful name.
+ * \tparam skills       A set of skills to be add to the expressive type.
+ *
+ * \details
+ *
+ * There are many cases in which interfaces use ambiguous parameters to
+ * that can be mixed up very easily.
+ * For example, if we consider an interface, that expects a window size and a number of maximal errors to search for
+ * possible hits in a region of interest, than both values might be given in form of an unsigned integer.
+ * The following snippet shows a perfectly well designed interface:
+ *
+ * ```cpp
+ * template <forward_range_concept fwd_rng_type>
+ * bool search(fwd_rng_type & rng, unsigned const w, unsigned const e)
+ * {
+ *  ...
+ * };
+ * ```
+ *
+ * The first parameter is the window size and the last parameter defines the error threshold.
+ * But, what happens if the user accidentally switches the `window_size` with the `error` parameter.
+ * In a bigger code base this mistake can be so subtle, that it becomes hard to figure out that the
+ * search interface was used in a wrong way. Also the compiler could not detect this.
```

Comment:
~~could not~~ cannot (you are using present tense in the paragraph)

Reply:
👍